src/bufr/BufrParser/Query/Data.h

```diff
@@ -0,0 +1,246 @@
+/*
+ * (C) Copyright 2023 NOAA/NWS/NCEP/EMC
+ *
+ * This software is licensed under the terms of the Apache Licence Version 2.0
+ * which can be obtained at http://www.apache.org/licenses/LICENSE-2.0.
+ */
+
+#pragma once
+
+#include <string>
+#include <vector>
+
+#include "Constants.h"
+
+namespace Ingester {
+namespace bufr {
+
+    /// \brief A class to hold data from a BUFR message. Unfortunatly this data can come either as
+    ///        long strings (strings larger than 8 bytes) or as octets (8 byte chunks). In order to
+    ///        handle both cases in a performant way we use a union to hold the data (std::variant
+    ///        is too slow). This means that we need to explicitly call the destructor and
+    ///        constructor for the union members. Many of the methods exposed on the class mimic
+    ///        those of std::vector.
+    struct Data
+    {
+        /// \brief A union to hold the data. Either a vector of doubles or a vector of strings.
+        union Value
+        {
+            std::vector<double> octets;
+            std::vector<std::string> strings;
+
+            Value() {}
+            ~Value() {}
+
+            /// \brief Explicitly call the constructor for the octet vector.
+            void initOctet()
+            {
+                new (&octets) std::vector<double>();
```

Comment:
@srherbener The destructor for the data `~Data()` object itself, manually calls the destructor of the active vector which should delete the dynamic content of the active vector. The meta data for the active vector itself is part of the union and should be destroyed when the union is destroyed.

Reply:
@srherbener The reason this works this way is because I used the "placement new" version of new (hence the funny syntax  [ex: `new (&octets) std::vector<double>();`]). Placement new lets you construct an object within pre-existing memory (the address of octets in this case). Since octets is declared as a by-value reference it will be managed accordingly.

Reply:
@rmclaren thanks for the explanation! Learn something "new" every day. I agree that the memory should be deallocated properly with the default destructors.

Reply:
@rmclaren I am just musing this could be just a std::vector<string> only. You have the flag below to denote if it's a double or actual string. You can write and read a string of 8 bytes for a double using std::stringstream or even old school file pointer to a memstreams.  Just an idea that might simplify the interface.

Reply:
@rmclaren You could just use both std::vector<double> and std::vector<std::string> and only use one rather that have a union.

Reply:
@PatNichols I am aware of that method for encoding data (actually you would need `std::vector<std::vector<double>>` or maybe `std::vector<std::string>` to accommodate strings > 8 bytes). I tried a lot of different ways to store the data and found I could get the best performance using this method. The Data object is used in very performance critical code so any time spent allocating objects (like all those std::vectors), re-interpreting or copying data around really eats your lunch (there can be a lot of data)...

Reply:
@PatNichols To your second point. Yes, I could always allocate vectors for both strings and octets (doubles) for each instance of `Data`, however, this is not optimal.  The ultimate reason is that BUFR is a very messy format. As you most likely know, a BUFR is a file that consists of a collection of "messages". Each "message" breaks into a set of what they call "subsets" (each one containing a complete data set usually anchored at some time). The issue is that different subsets in a file don't necessarily have the same format for a data field, but can vary in different ways. For example, a field in one subset may have different dimensions as a field in a different subset. The differences between message subsets must be resolved before the data can be assembled into a usable dataset. The resolution can not happen until the data has all been collected.  Unfortunately its not as simple as just pushing the data into an accumulator and just returning it... 

In the context of is project each message::subset data is accumulated into a "frame" (all the data we want for that message subset, usually a lot more than one field). There could be hundreds of thousands frames 
(there is no upper limit really).  Each frame (a table of stuff) consists of a collection of `Data` instances for each kind of data collected. There could be millions of these things... So its really important that these things are a lightweight as possible. Adding just an extra allocation for a std::vector can have a surprising effect on the overall performance.

**FYI:** it is easier to see whats going on in the `feature/query_new_results` branch.

**Note:** I spent a lot of time trying things and running them through Apple Instruments to figure out the most efficient way to do things. Unfortunately I had to make the code a little more complicated to get the performance we need.